### PR TITLE
Fix CLOSE_STDIN sentinel

### PR DESCRIPTION
**Pull request by blueyed**
This makes it / it's repr more meaningful in docs, and should have been
`object()` anyway.

TODO:

- [x] after https://github.com/pytest-dev/pytest/pull/6872

### CODE REVIEW

File: src/_pytest/pytester.py [hunk 1]
```diff
@@ -534,7 +534,8 @@ class Testdir:
 
     __test__ = False
 
-    CLOSE_STDIN = object
+    class CLOSE_STDIN:
+        pass
```
Comment (bluetech) [inline src/_pytest/pytester.py]:
```suggestion
        __slots__ = ()
```

Comment (bluetech) [reply src/_pytest/pytester.py]: Hmm given that you use the class itself as the sentinel and not an instance of it, this suggestion doesn't matter much...

**Comment by blueyed** [comment]
> maybe handrolling a version of `unittest.mock.sentinel`

I'm not opposed to it.

Given that this needs to be rebased anyway suggestions are welcome.
OTOH we're not using it that much currently, and therefore I think it's fine to go with it for now.

File: src/_pytest/pytester.py (hunk 1 above)
Comment (graingert) [reply src/_pytest/pytester.py]: an enum instance is usually pretty good

```python3
class CloseStdinType(enum.Enum):
    CLOST_STDIN = enum.auto()
```

it's great for type annotations too

Comment (graingert) [reply src/_pytest/pytester.py]: and it's got that None/NoneType feel

Comment (graingert) [reply src/_pytest/pytester.py]: @blueyed you can do stuff like:
```
def popen(self, ..., stdin=CLOSE_STDIN: Union[CloseStdinType,bytes,int]): -> ...
```

Comment (bluetech) [reply src/_pytest/pytester.py]: @graingert Right, this is the best current way to do it in type-compatible way: https://github.com/python/typing/issues/689. Also what I opted for here: https://github.com/bluetech/pytest/commit/bcd19fac5b30882e106ef1e3dbad4bb7edb8b303#diff-4ec37b1ad5b7d41fdac56c907d08dde1R46-R52

Comment (blueyed) [reply src/_pytest/pytester.py]: What's wrong with `stdin: Optional[Union[IO[str], "EncodedFile", "DontReadFromInput", Type[CLOSE_STDIN]]] = CLOSE_STDIN,`?
It might be not as strict really, but serves the purpose in general, and has less/no boilerplate.

Comment (bluetech) [reply src/_pytest/pytester.py]: Nothing wrong really, just that using a `Type[...]` as a sentinel is odd/surprising. I'd go with @graingert's suggestion, which is the ["official"](https://www.python.org/dev/peps/pep-0484/#support-for-singleton-types-in-unions) way to do it.

That said if you prefer the Type that's fine.

File: src/_pytest/pytester.py [hunk 2]
```diff
@@ -513,7 +519,7 @@ def restore(self) -> None:
         sys.path[:], sys.meta_path[:] = self.__saved
 
 
-class Testdir:
+class Testdir(Generic[AnyStr]):
```
Comment (bluetech) [inline src/_pytest/pytester.py]: I don't see `AnyStr` mentioned elsewhere, why did you need this generic?

File: src/_pytest/pytester.py [hunk 3]
```diff
@@ -1089,14 +1096,42 @@ def collect_by_name(
...
         return None
 
+    @overload
     def popen(
         self,
         cmdargs,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.PIPE,
-        stdin=CLOSE_STDIN,
+        stdout: Optional[Union[int, IO]] = subprocess.PIPE,
```
Comment (bluetech) [inline src/_pytest/pytester.py]: Is it specifically `IO[str]` or `IO[bytes]`?

I see that typeshed punts on this: https://github.com/python/typeshed/blob/d8b081130d3e3a7fde1a49ac787d304f5c4c1966/stdlib/3/subprocess.pyi#L13-L27

File: src/_pytest/pytester.py [hunk 4]
```diff
@@ -1089,14 +1096,42 @@ def collect_by_name(
...
+        stderr: Optional[Union[int, IO]] = subprocess.PIPE,
+        stdin: Optional[Union[CloseStdinType, bytes, int, IO]] = CLOSE_STDIN,
```
Comment (bluetech) [inline src/_pytest/pytester.py]: Does `Literal[CLOSE_STDIN]` work? If yes, that'd be a little nicer. (I think it is planned to work, but I don't remember if it does already).

File: src/_pytest/pytester.py (hunk 3 above)
Comment (blueyed) [reply src/_pytest/pytester.py]: I've used `AnyStr` here initially (but forgot to clean it up).
I've figured `IO` is the same as `IO[Any]`.  Do you prefer the latter?
We basically have the `_FILE` from typeshed here: `Union[None, int, IO[Any]]`

Comment (bluetech) [reply src/_pytest/pytester.py]: I'd leave it with the explicit `Any` (like you did), maybe can be filled with something precise in the future.

Comment (blueyed) [reply src/_pytest/pytester.py]: It can be `str` or `bytes`.  typeshed binds it according to overloads.

**Comment by nicoddemus** [comment]
Closing for now.
